### DISCUSSION TITLE
General recursion on types and expressions

### DISCUSSION THREAD

**Discussion by fbrausse** [esbmc/esbmc#1208] [Ideas]
In a lot of places inside ESBMC we have adhoc recursions finding out about properties of types and expressions or that modify those in a particular way. Sometimes both. I'm opening this discussion in order to find a way to unify these recursions. The reason is that the adhoc variants do like to miss subtleties like these:
- an array type requires a recursion into its array size (if exists) as well as into its subtype
- struct and union types behave the same wrt. to recursion: recurse into all its members, there usually is no need to special-case on just one of them
- pointer types are often overlooked, they need recursion into their subtype as well
- symbolic types are not handled
- all expressions are typed, they need recursion also into their types

In addition to this (probably incomplete) list of things to watch out for, a more general property of our data structures is almost always¹ overlooked: they are DAGs (directed acyclic graphs), not trees. By treating them as trees with adhoc recursion we're not only duplicating work for equal nodes inside the dag, but potentially also explode the size of the DAG. Example: symbols `x,y` both of pointer type to `t`, expression `e` denoting a function call `g(x, y)`. Suppose want to adjust `t` to `f(t)`. Then, by adhoc-recursing into both operands of `g` and treating the `x,y` things independently, we're applying `f(t)` twice, potentially even transforming the DAG into a tree.

So, in order to address this, I'm proposing to introduce a general recursion scheme that takes these things into account. I'll try to introduce in another PR, but that will take time. Meanwhile, this discussion is about things I may have missed and proposals how to potentially get such a general recursion scheme to work.

¹ the notable exception that comes to mind is `thrash_type_symbols()`; there may be other instances, but they are rare

**Comment by mikhailramalho** [comment]
My understanding is that whenever we are handling types in any way, we should call ns.lookup() first, but we were never consistent about that.

**Reply by fbrausse** [reply]
That's another issue we should decide once and for all: is a single-step ns.lookup() enough or do we expect multiple levels of symbolic types that need ns.follow()?

**Reply by mikhailramalho** [reply]
I think only one should be needed

**Comment by lucasccordeiro** [comment]
@fbrausse: I'm totally in favor to introduce a general recursion scheme.

**Comment by fbrausse** [comment]
To illustrate what I have in mind wrt. "recursion on DAGs", here is an example how such a scheme could look like. Consider only a single `term` type (instead of our expr2tc and type2tc) and pointers to terms called `term_ptr`. Think of `term_ptr` as a pointer to a `term` node inside a graph (or DAG) structure, i.e., they are unique wrt. to `operator==` and their hash and they allow access to their children (via outgoing edges) and, if needed, to their parents as well (via incoming edges).

Then we can build sets of terms (term_set) and maps from terms to some type R (term_map<R>). In another project, I've used the following code to implement the DAG-like recursion mentioned initially like so:
```c++
/* term_set is just a set of terms */
/* term_map<R> is a map from terms to R */

template <typename R, typename F, typename... Args>
static inline R with_unique_term_rec(term_ptr p, F &&f, term_map<R> &map, Args &&... args) /* internal recursion with term_map */
{
	auto it = map.find(p);
	if (it != end(map))
		return it->second;
	return map.emplace(p, f(p, [&] <typename... Brgs>
	                           (term_ptr q, Brgs &&... brgs) {
		return with_unique_term_rec(q, f, map, std::forward<Brgs>(brgs)...);
	}, std::forward<Args>(args)...)).first->second;
}

template <typename F, typename... Args>
static inline void with_unique_term_rec(term_ptr p, F &&f, term_set &set, Args &&... args) /* internal recursion with term_set */
{
	auto [it,ins] = set.emplace(p);
	if (!ins)
		return;
	f(p, [&] <typename... Brgs> (term_ptr q, Brgs &&... brgs) {
		return with_unique_term_rec(q, f, set, std::forward<Brgs>(brgs)...);
	}, std::forward<Args>(args)...);
}

template <typename R = void, typename F, typename... Args>
static inline auto with_unique_term(term_ptr p, F &&f, Args &&... args) /* entry point of the recursion */
{
	if constexpr (std::is_same_v<std::remove_cvref_t<R>,void>) {
		term_set set;
		with_unique_term_rec(p, f, set, std::forward<Args>(args)...);
		return set;
	} else {
		term_map<R> map;
		with_unique_term_rec(p, f, map, std::forward<Args>(args)...);
		return map;
	}
}
```
The entry point `with_unique_term()` can be called like so: `term_map<int> map = with_unique_term<int>(ptr, my_transformation, dummy);` where `float dummy` is just a parameter for demonstration. The `my_transformation` function could have this signature:
```c++
template <typename Rec>
int /* or any other type, including void */
my_transformation(term_ptr p, Rec &&rec, float dummy); /* calls e.g. rec(p->get_child(0), dummy+1.5f); */
```
This would ensure that `my_transformation()` is run at most once on each node of the DAG and produces only a single result of that node. The return value `map` is a `term_map<int>` here, because `my_transformation()` returns an `int`. It could just as well return a `term_ptr` or anything else. That map is *the result* of the recursion. In some sense, I'd like to move away from the view of manipulating terms via a recursion, and rather see it as a transformation from a DAG to something else.

This approach takes a somewhat functional view where the result only depends on the term_ptr the function is operated on. In particular, the existing terms inside the DAG are immutable. If one wants to change `ptr->child[0]` to something else, a new node in the dag has to be created and necessarily a `term_ptr` needs to be part of the return value... The code above could be extended to use a `map<std::tuple<term_ptr,float>,int>` as well, to make the recursion also dependent on `dummy`. For the other project I didn't need this generality.

I'd be interested in opinions on how feasible this is for ESBMC, where an explicit DAG structure on expr2tc/type2tc could fit in and also potential problems you might see with such an approach.
